### DISCUSSION TITLE
@RetryableTopic to include and exclude expection

### DISCUSSION THREAD

**Discussion by ajitavrb** [spring-projects/spring-kafka#2926] [Q&A]
Hi, I am trying to use @Retryable topic annotation, however I am unable include both `include` and `exclude` in the exception list to retry on and not to retry on respectively.

```
    @RetryableTopic(
            backoff = @Backoff(delay = 30000L),
            attempts = "2",
            autoCreateTopics = "false",
            sameIntervalTopicReuseStrategy = SameIntervalTopicReuseStrategy.SINGLE_TOPIC,
            dltStrategy= DltStrategy.FAIL_ON_ERROR,
            autoStartDltHandler = "true",
            include = CustomRetryException.class,
            exclude = {SQLException.class})
    @KafkaListener(topics = EVENT_TOPIC, groupId = "in-test-consumer")
    public void consumeEvent(@Payload ConsumerRecord<Key, Value> message) {
        log.info("Event received {}", message);
    }
```

```
Caused by: java.lang.IllegalArgumentException: Please use only retryOn() or only notRetryOn()
```

Is there a possible way to configure an exception so that it will never be retried?

**Comment by artembilan** [comment]
It is by design from the `BinaryExceptionClassifier` API.
So, only `include` - therefore the rest exception are going to be non-retried.
Or only `exclude` - therefore only these exceptions are non-retried, the rest are.

You need to design your decision this or other way: if you have more exceptions to exclude, then just add them respectively.
